File: lib/builtins/flutter/widgets/safeArea.dart
import 'package:flua/hydroState.dart';
import 'package:flua/vm/context.dart';
import 'package:flua/builtins/flutter/syntheticBox.dart';
import 'package:flua/vm/table.dart';
import 'package:flutter/material.dart';

loadSafeArea({@required HydroState luaState, @required HydroTable table}) {
  table["safeArea"] = makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      SafeArea(
        top: args[0]["top"],
        bottom: args[0]["bottom"],
        left: args[0]["left"],
        right: args[0]["right"],
        maintainBottomViewPadding: args[0]["maintainBottomViewPadding"],
        minimum: args[0]["minimum"],
        child: maybeUnwrapAndBuildArgument<Widget>(args[0]["child"],
            parentState: luaState),
      )
    ];
  });
}

File: runtime/flutter/widgets/safeArea.ts
import { Type } from "../../dart/core/type";
import { RuntimeBaseClass } from "../../runtimeBaseClass";
import { EdgeInsets } from "./../painting/edgeInsets";
import { Widget } from "./../widget";
import { StatelessWidget } from "./statelessWidget";

export interface SafeAreaProps {
    left?: boolean | undefined;
    top?: boolean | undefined;
    right?: boolean | undefined;
    bottom?: boolean | undefined;
    minimum?: EdgeInsets | undefined;
    maintainBottomViewPadding?: boolean | undefined;
    child: Widget;
}

declare const flutter: {
    widgets: {
        safeArea: (this: void, props: SafeAreaProps) => SafeArea;
    };
};

export class SafeArea extends StatelessWidget implements RuntimeBaseClass {
    public readonly internalRuntimeType = new Type(SafeArea);
    public props: SafeAreaProps;
    public constructor(props: SafeAreaProps) {
        super();
        this.props = props;

        if (this.props.left === undefined) {
            this.props.left = true;
        }
        if (this.props.top === undefined) {
            this.props.top = true;
        }
        if (this.props.right === undefined) {
            this.props.right = true;
        }
        if (this.props.bottom === undefined) {
            this.props.bottom = true;
        }
        if (this.props.minimum === undefined) {
            this.props.minimum = EdgeInsets.zero();
        }
        if (this.props.maintainBottomViewPadding === undefined) {
            this.props.maintainBottomViewPadding = false;
        }
    }

    public build(): Widget {
        return flutter.widgets.safeArea(this.props);
    }
}
